### PR TITLE
Allowing remote_id for user group related migrations

### PR DESCRIPTION
**Pull request by andesk**
As in our project we do not want to rely on autoincrement id's related to user group migrations (e.g. as resorting might break migrations), we enhanced the existing migration possibilities for user groups to allow usage of remote_id for:
- user group's "parent_group_id": if string provided, it is considered referencing a user group's remote_id instead of it's id
- allowing matching on "content_remote_id" for user group update/delete 
- allowing usage of user group remote_id in user's "groups" when creating/updating

I hope this PR work is acceptable, otherwise I am open to improvement ideas and suggestions. Did not find any unit tests to enhance or adapt, tested running several migrations using the new possibilities though.

### CODE REVIEW

File: Core/Executor/UserGroupManager.php [hunk 1]
```diff
@@ -31,14 +31,17 @@ protected function create()
 
         $parentGroupId = $this->dsl['parent_group_id'];
         $parentGroupId = $this->referenceResolver->resolveReference($parentGroupId);
-
-        $parentGroup = $userService->loadUserGroup($parentGroupId);
+        $parentGroup = $this->userGroupMatcher->matchOneByKey($parentGroupId);
```
Comment (andesk) [inline Core/Executor/UserGroupManager.php]: Using matcher instead, copied that approach from existing places like ContentManager.

File: Core/Executor/UserGroupManager.php [hunk 2]
```diff
@@ -69,7 +72,7 @@ protected function create()
      */
     protected function update()
     {
-        $userGroupCollection = $this->matchUserGroups('delete');
+        $userGroupCollection = $this->matchUserGroups('update');
```
Comment (andesk) [inline Core/Executor/UserGroupManager.php]: Little "typo" it seems.

File: Core/Executor/UserManager.php [hunk 3]
```diff
@@ -111,25 +115,29 @@ protected function update()
 
                 $assignedGroups = $userService->loadUserGroupsOfUser($user);
 
+                $targetGroupIds = [];
                 // Assigning new groups to the user
                 foreach ($this->dsl['groups'] as $groupToAssignId) {
+                    $groupId = $this->referenceResolver->resolveReference($groupToAssignId);
+                    $groupToAssign = $this->userGroupMatcher->matchOneByKey($groupId);
```
Comment (andesk) [inline Core/Executor/UserManager.php]: This might lead to a slightly  slower performance as before not every group was loaded from storage, but otherwise we can not cover all required checks below...

File: Core/Matcher/ContentMatcher.php [hunk 4]
```diff
@@ -56,7 +56,7 @@ public function matchContent(array $conditions)
 
             // BC support
             if ($key == 'content_type') {
-                if (is_int($key) || ctype_digit($key)) {
+                if (ctype_digit($key)) {
```
Comment (andesk) [inline Core/Matcher/ContentMatcher.php]: As we checked the $key 's value one line higher for being a string, this whole check is useless, no?

File: Core/Matcher/UserGroupMatcher.php [hunk 5]
```diff
@@ -75,4 +88,25 @@ protected function findUserGroupsById(array $userGroupIds)
...
+     * @return UserGroup[]
+     */
+    protected function findUserGroupsByContentRemoteIds(array $remoteContentIds)
+    {
+        $userGroups = [];
+
+        foreach ($remoteContentIds as $remoteContentId) {
+            // return unique contents
+
+            // user service does not provide a method to load user groups via remote_id, but as user groups are content...
```
Comment (andesk) [inline Core/Matcher/UserGroupMatcher.php]: See this comment. As the user service does not provide a method to load user group by remote_id, but a user group is stored as a content which supports remote id of course, I added this "hack" to first fetch the content by remote_id and then fetch the user group by the content's id. Again, slight performance impact for that use case, but at least we cover the use case :)

File: Core/Executor/UserGroupManager.php (hunk 1 above)
Comment (gggeek) [inline Core/Executor/UserGroupManager.php]: This change is not harmful, but it does not add a lot of value either, as we still have to use the userService in this method, and it is a bit slower...

File: Core/Matcher/ContentMatcher.php (hunk 4 above)
Comment (gggeek) [inline Core/Matcher/ContentMatcher.php]: I think there is an error here: we should probably be checking $values...

File: Core/Executor/UserManager.php (hunk 3 above)
Comment (gggeek) [inline Core/Executor/UserManager.php]: I presume that this means you could identify the groups to assign to a user by ref as well as remote id, correct?

File: Core/Executor/UserGroupManager.php (hunk 1 above)
Comment (andesk) [inline Core/Executor/UserGroupManager.php]: Sure, but the matcher can use different formats of Ids, while the User Services expects and id here, no?

File: Core/Executor/UserManager.php (hunk 3 above)
Comment (andesk) [inline Core/Executor/UserManager.php]: Exactly, that's what I meant with "allowing usage of user group remote_id in user's "groups" when creating/updating"

File: Core/Matcher/ContentMatcher.php (hunk 4 above)
Comment (andesk) [inline Core/Matcher/ContentMatcher.php]: Not part of this PR though, are you raising/checking this independently?

**Comment by andesk** [comment]
Oh great, already merged? So it'll be part of the next release?

File: Core/Executor/UserGroupManager.php (hunk 1 above)
Comment (gggeek) [inline Core/Executor/UserGroupManager.php]: Indeed. I just saw that you had extended it after leaving that comment

**Comment by gggeek** [comment]
Yes.
I planned to release today, but the other changes which got merged warrant some further dev work. I now hope to have the release out by the end of next week.
